Add a few functions & POSIX error codes to Windows API

mikehoyle opened the pull request:
This change adds a few simple additional functions to the windows apis, as well as some additional error codes.

Tests and style pass, as per contributing guidelines.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @gnzlbg (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

retep998 commented:
`ucrt\errno.h` does have those error constants, and notably they are *different* than the WSA constants of the same name. For example `EADDRINUSE` from ucrt has a value of 100 but `EADDRINUSE` from winsock2 has a value of 10048. Obviously a massive footgun but it is what it is.

`time` and `_lseeki64` are both from ucrt as well. Notably however `time` is actually an inline wrapper around `_time64`. The signatures look correct otherwise.

UdamSaini commented:
Inside winsock.h, they say to not use EADDRINUSE and to use WSAEADDRINUSE due to the conflict for socket related errors. 
/*
* Windows Sockets errors redefined as regular Berkeley error constants.
* These are commented out in Windows NT to avoid conflicts with errno.h.
* Use the WSA constants instead.
*/
#if 0
#define EWOULDBLOCK  WSAEWOULDBLOCK
......

I think we should probably use the errno constants as that is what windows uses, and have the WSA constants be separate or in the win32 library? I wonder if we should add a comment to that effect, similar to the windows comment.

mikehoyle commented:
In my most recent commit I updated the time funtion to link to _time64 (as time is just a wrapper around that anyways). It also seems that the gnu toolchain uses a different (probably outdated) version of errno.h that doesn't contain EOTHER, so I have moved that specific constant to the msvc-specific file.

gnzlbg commented on src/windows/mod.rs:
```diff
@@ -183,6 +183,48 @@ pub const ENOTEMPTY: ::c_int = 41;
 pub const STRUNCATE: ::c_int = 80;
 
+// POSIX Supplement (from errno.h)
+pub const EADDRINUSE: ::c_int = 100;
```
Per the comment below from @UdamSaini , does it make sense to add this one here ? (or maybe we can add it as `#[deprecated]` with the same warning as the windows headers?

retep998 replied:
Those warnings are on the winsock definitions that just redefine to the WSA constants, not the errno definitions.

gnzlbg commented:
@bors: r+

bors commented:
:hourglass: Testing commit 3ecdafbde86fc01cd3f37c9327497ed0da8ab743 with merge fe7ccff8981ea0ec113cc9e4cb2180f8b5d381a9...
<!-- homu: {"type":"BuildStarted","head_sha":"3ecdafbde86fc01cd3f37c9327497ed0da8ab743","merge_sha":"fe7ccff8981ea0ec113cc9e4cb2180f8b5d381a9"} -->

bors commented:
:sunny: Test successful - [checks-cirrus-freebsd-10](https://cirrus-ci.com/task/5643261321936896), [checks-cirrus-freebsd-11](https://cirrus-ci.com/task/5080311368515584), [checks-cirrus-freebsd-12](https://cirrus-ci.com/task/6206211275358208), [status-azure](https://dev.azure.com/rust-lang2/4a14b271-efdb-4310-b967-ee2fa53207f6/_build/results?buildId=713)
Approved by: gnzlbg
Pushing fe7ccff8981ea0ec113cc9e4cb2180f8b5d381a9 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"gnzlbg","base_ref":"master","builders":{"status-azure":"https://dev.azure.com/rust-lang2/4a14b271-efdb-4310-b967-ee2fa53207f6/_build/results?buildId=713","checks-cirrus-freebsd-10":"https://cirrus-ci.com/task/5643261321936896","checks-cirrus-freebsd-11":"https://cirrus-ci.com/task/5080311368515584","checks-cirrus-freebsd-12":"https://cirrus-ci.com/task/6206211275358208"},"merge_sha":"fe7ccff8981ea0ec113cc9e4cb2180f8b5d381a9"} -->
